src/frontend/src/utils/findWebAuthnRpId.ts

```diff
@@ -0,0 +1,107 @@
+ * @returns {DeviceData[]} The list of devices registered for the domain.
+ */
+const getDevicesForDomain = (
+  devices: DeviceData[],
+  domain: string
+): DeviceData[] =>
+  devices.filter((d) => {
+    if (d.origin.length === 0)
+      return domain === getTopAndSecondaryLevelDomain(defaultDomain);
+    return d.origin.some((o) => getTopAndSecondaryLevelDomain(o) === domain);
+  });
+
+/**
+ * Returns the domain to use as the RP ID for WebAuthn registration.
+ *
+ * The algorithm is as follows:
+ * 1. If there is a device registered for the current domain, return undefined.
+ * 2. If there is no device registered for the current domain, check if there is a device registered for one of the preferred domains.
+ *    If there is, return the first preferred domain that has a device registered.
+ * 3. If there is no device registered for the current domain and none of the preferred domains.
+ *    Raise an error because the devices should be registered in one of the preferred domains.
+ *
+ * @param currentUrl - The current URL of the page.
+ * @param devices - The list of devices registered for the user.
+ * @param preferredDomains - Optional list of domains in order or preference to use as the RP ID.
+ * @returns {string | undefined} The RP ID to use for WebAuthn registration.
+ * `undefined` when the RP ID is the same as the current domain and is not needed.
+ *
+ * @throws {Error} If devices are not registered for any of the preferred domains.
+ * @throws {Error} If no devices exist or the current domain is invalid.
+ * @throws {Error} If the current domain is invalid.
+ */
+export const findWebAuthnRpId = (
+  currentUrl: string,
+  devices: DeviceData[],
+  preferredDomains: string[] = ["ic0.app", "internetcomputer.org", "icp0.io"]
+): string | undefined => {
+  try {
+    const currentDomain = getTopAndSecondaryLevelDomain(currentUrl);
+
+    if (devices.length === 0) {
+      throw new Error(
+        "Not possible. Every registered user has at least one device."
+      );
+    }
+
+    const getFirstDomain = (devices: DeviceData[]): string => {
+      if (devices[0] === undefined) {
+        throw new Error(
+          "Not possible. Call this function only if devices exist."
+        );
+      }
+      return devices[0].origin[0] ?? defaultDomain;
+    };
+
+    // Try current domain first if devices exist
+    if (getDevicesForDomain(devices, currentDomain).length > 0) {
+      return undefined;
+    }
+
+    // Check based on the order of preferred domains if there is no device with the current domain.
+    for (const domain of preferredDomains) {
+      const devicesForDomain = getDevicesForDomain(devices, domain);
+      if (devicesForDomain.length > 0) {
+        return getFirstDomain(devicesForDomain);
+      }
+    }
+
+    throw new Error(
+      "Not possible. Devices must be registered for at least one of the following domains: ic0.app, internetcomputer.org, icp0.io"
+    );
+  } catch (error) {
+    console.error(error);
```

Comment:
Instead of wrapping the whole `findWebAuthnRpId` function in a try/catch just to console log any errors, maybe the caller of `findWebAuthnRpId` can console log the error instead to avoid the need for the try/catch?

Reply:
Yes, I initially added it wanting to return something different. But then I decided against it and the try/catch was left. I'll remove it.